### PR TITLE
[source-postgres] Disable incremental sync for view in Postgres xmin replication mode

### PR DESCRIPTION
**Pull request by theyueli**
## What

Fixes https://github.com/airbytehq/airbyte-internal-issues/issues/8559
Fixes https://github.com/airbytehq/oncall/issues/5581

This patch disables incremental sync when a stream is found to be a view. This is done only for Postgres sources configured in xmin mode.

## How

Adding a check in discovery() to only allow `FULL_REFRESH` sync mode.

## User Impact
In the UI, user will only see full refresh sync option for view. See the picture below.
Screenshot 2024-07-18 at 6 09 50 PM

## Can this PR be safely reverted and rolled back?

- [X ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java [hunk 1]
```diff
@@ -75,6 +75,19 @@ public static AirbyteStream setIncrementalToSourceDefined(final AirbyteStream st
...
   }
 
+  public static boolean isStreamAView(final JdbcDatabase database, final AirbyteStream stream) throws SQLException {
+    final String query = "SELECT COUNT(*) FROM pg_catalog.pg_views WHERE schemaname = ? AND viewname = ?";
+    final List<JsonNode> result = database.queryJsons(query, stream.getNamespace(), stream.getName());
+    return result.get(0).get("count").asInt() > 0;
+  }
+
+  public static AirbyteStream setFullRefreshToView(final JdbcDatabase database, final AirbyteStream stream) throws SQLException {
+    if (isStreamAView(database, stream)) {
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:85]: I think this means we will run this query N times, once for all N streams.  Could we load up the list of streams once instead?

Comment (theyueli) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java]: Great idea! I just pushed a commit that implements this idea. Please take another look, thanks!

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java [hunk 2]
```diff
@@ -61,6 +58,42 @@ public static AirbyteStream overrideSyncModes(final AirbyteStream stream) {
...
+   * This method is used for CDC sync in order to overwrite sync modes for cursor fields cause cdc use
+   * another cursor logic. Used in discovery when xmin is configured.
+   *
+   * @param stream - airbyte stream
+   * @param viewsBySchema - view names by schema
+   * @return will return list of sync modes where view streams will have only FULL_REFRESH sync mode
+   */
+  public static AirbyteStream overrideSyncModes(final AirbyteStream stream, Map<String, List<String>> viewsBySchema) {
+    if (isStreamAView(viewsBySchema, stream)) {
+      return stream.withSupportedSyncModes(Lists.newArrayList(SyncMode.FULL_REFRESH));
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:92]: I'm not sure, but I thought in an xmin sync, full refreshes were still incremental... and the therefore still dependent on a usable cursor?

Comment (theyueli) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:92]: I think this is for the cursor-based mode for xmin sync; 

maybe I'm wrong, but from what I tested, it does not require specifying a cursor in xmin  (see picture below)
Screenshot 2024-07-18 at 7 30 15 PM

Comment (akashkulk) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:92]: The full refreshes are still incremental (based on CTID), but I didn't think a cursor needed to be explicitly set? So I think fixing the sync modes might work 

cc : @xiaohansong who might know this

Comment (xiaohansong) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:92]: yes - cursor is not needed for full refresh. We should support full refresh anyway, for views they would be non resumable.

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java [hunk 3]
```diff
@@ -61,6 +58,42 @@ public static AirbyteStream overrideSyncModes(final AirbyteStream stream) {
     return stream.withSupportedSyncModes(Lists.newArrayList(SyncMode.FULL_REFRESH, SyncMode.INCREMENTAL));
   }
 
+  public static List<String> getViewsInSchema(final JdbcDatabase database, final String schema) throws SQLException {
+    final String query = "SELECT viewname FROM pg_catalog.pg_views WHERE schemaname = ?";
```
Comment (xiaohansong) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:62]: I think this should work, but the existing solution is to check for filenode of the stream. Maybe we should consider consolidate them? 

See `supportResumableFullRefresh` in `PostgresSource.java`

**Review by xiaohansong — APPROVED** [review]

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java (hunk 3 above)
Comment (theyueli) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCatalogHelper.java:62]: @xiaohansong and I had a discussion and we decide to keep the current implementation. We will also file another ticket to track a potential improvement we could do on this file node function that issues a db query to each stream.

**Comment by taherajnaFA** [comment]
If I understand this change correctly, we can no longer sync a view in an incremental mode, right?
Currently, we have a use case where we want to filter the base table of our Postgres source within a certain time frame, essentially we do not wish to sync old data into our warehouse. 
So the way we have achieved this is by creating a view on top of the base table with the time filter applied, also we have selected the xmin column while creating the view.
To Airbyte, up until now this would just behave as a raw table, all we had to do was unselect the xmin column from the schema while setting up the connection and Airbyte would sync everything incrementally to our destination.
After this change, i assume that this would no longer work.

I think instead of removing the option out right, this change should be implemented in a different way.

**Comment by evantahler** [comment]
> If I understand this change correctly, we can no longer sync a view in an incremental mode, right?

Only when using the xmin replication mode

**Comment by taherajnaFA** [comment]
@evantahler precisely, that's what our usecase is. We just have a filter on the start date of the source table in our view, everything stays as is.

**Pull request merged**